---
title: "![](orthomapper-icon.png) Orthomapper: Mapping Orthologs"
output: 
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{Finding orthologous genes}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(pander)
library(orthomapper)
library(org.Hs.eg.db)
library(org.Mm.eg.db)
```

# Basic usage

## Quick start

```{r eval=FALSE}
library(orthomapper)
## get all orthologs between mouse and rat
mouserat <- orthologs(10090, 10116)

## get the SYMBOL for gene 52024, infer taxon automatically
entrez_annotate(52024)

## get more information on the gene ID 52024
entrez_annotate(52024, taxon=10090, c("SYMBOL", "GENENAME"))

## get all orthologs of the gene ID 52024
ankrd22.o <- orthologs_all(52024)

## map a list of genes to orthologs from another taxon
orthomap(c(229898, 52024), 10090, 9606)

## find a gene symbol by similarity search
similar_symbol("GBP5", 10090)

## get all the species in the databse
species_info()

## what is taxon 9606?
species_info(9606)

## what is the taxon ID of the mouse?
species_search("mouse")
```


## Orthomapper and taxonomy

To use orthomapper, you need to know the Taxonomy IDs (from
[NCBI](https://www.ncbi.nlm.nih.gov/taxonomy)) of the species you are
working on. Some of the more popular Taxonomy IDs are given below in Table
1.

To get all orthologous gene pairs from two taxons (termed "source" and
"target"), you simply can run the following function:

```{r}
mouserat <- orthologs(10090, 10116)
dim(mouserat)
colnames(mouserat)
head(mouserat)
```

We can also specify a vector with gene IDs and ask for specific orthologs
to that genes:

```{r}
## map Gbp5 and Ankrd22 from mouse to human 
orthomap(c(229898, 52024), 10090, 9606)
```

To get all orthologs of the gene 52024 (Ankrd22), run:

```{r}
ankrd22.o <- orthologs_all(52024)
head(ankrd22.o[[1]])
```

`orthologs_all` returns a list, since you can submit more than one gene at
once.


**Table 1.** Species for which organism annotation is available.

```{r,results="asis",echo=FALSE}
.dbi <- speciesDBITable()
pandoc.table(.dbi[ order(.dbi$species), ], justify="left")
```




## Mapping between Entrez IDs and other types of IDs

As you can see, the orthomapper DB is based on Entrez gene IDs. 
Fortunately, orthomapper makes translating the Entrez IDs relatively
painless. In the table above, you will see that for several species,
orthomapper is aware which annotation package should be used. If asked to
do so, it attempts to load the respective package and return the requested
columns such as "SYMBOL" or "GENENAME":

```{r}
# infer taxon automatically
entrez_annotate(52024)

# specify taxon and columns
entrez_annotate(52024, taxon=10090, c("SYMBOL", "GENENAME"))
```

If you do not specify the taxon, orthomapper uses the function
`find_taxon()` to query the orthology database about the taxon
corresponding to the gene ID. It will fail if the first gene of the genes
argument is not in the database.

You can also provide the name of an annotation package to use with `entrez_annotate`:

```{r eval=FALSE}
entrez_annotate(52024, orgdb="org.Mm.eg.db")
```

Orthomapper will then try to load the package and search for the respective
columns. By default, the query is assumed to be ENTREZID (`keytype`
parameter), and
`entrez_annotate` attempts to retrieve the SYMBOL column. If these columns
are not in the annotation db, or that is not what you want, you need to
specify the columns using the `keytype` and `column` parameters:


```{r eval=FALSE}
entrez_annotate("myCustomID", 
  keytype="CustomID", column="Description", orgdb="myCustomOrg.db")
```


## Taxon IDs and species names

For the 300+ species in the orthology database species information from
NCBI is directly provided. For each taxon ID, you can retrieve the
scientific species name, common names:

```{r}
## by default, lineage information is not shown
species_info()[1:4,]

## what is the species 10090?
species_info(10090)

## get the lineage for 10090
species_info(10090, show.lineage=TRUE)$lineage
```

You can also quickly get the taxonomy ID for a species by searching through
the fields of the species database with a pattern:

```{r}
## find the mouse
species_search("mouse")

## find all zebras ;-)
species_search("zebra")
```

# Gory details

## Source of the data

The package `orthomapper` is based on a file containing information about
orthologous genes distributed by NCBI, `gene_orthologs.gz` downloaded from
[https://ftp.ncbi.nlm.nih.gov/gene/DATA/](https://ftp.ncbi.nlm.nih.gov/gene/DATA/). 
This file has been converted into SQLite database. Note that the file does
not contain all ortholog pairs; rather, it contains ortholog pairs linking
to "hub" species (such as humans), so in order to find all orthologs
between two species you need to first identify all orthologs between these
species and a hub species, and then match the respective pairs.

For example, a mouse gene M1 might be orthologous to a rat gene R1, but
this information is not contained directly in the ortholog database.
Rather, R1 and M1 both link to a human gene H1, so the relationship between
R1 and M1 must be inferred from the relationships R1-H1-M1. This is the
point of the `orthomapper` package.

## Database structure

The Orthomapper database resides in `extdata/orthologDB.sqlite`. You can
find this file in the R installation with

```{r}
system.file("extdata", "orthologDB.sqlite", package="orthomapper")
```

The database is an SQLite file with currently three tables:

 * orthologs: `TaxID1, GeneID1, TaxID2, GeneID2`
 * species: `TaxID, name, common_name, lineage`
 * meta: additional meta information

